README.md
<div align="center">
    <a href="#"><img src="logo-linna-128.png" alt="Linna Logo"></a>
</div>

<br/>

<div align="center">
    <a href="#"><img src="logo-app.png" alt="Linna dotenv Logo"></a>
</div>

<br/>

<div align="center">

[![Tests](https://github.com/linna/app/actions/workflows/tests.yml/badge.svg)](https://github.com/linna/app/actions/workflows/tests.yml)
[![PDS Skeleton](https://img.shields.io/badge/pds-skeleton-blue.svg?style=flat)](https://github.com/php-pds/skeleton)
[![PHP 8.1](https://img.shields.io/badge/PHP-8.1-8892BF.svg)](http://php.net)

</div>

# About
Application Skeleton for Linna framework

## Actual stable environment
* app 0.16.0
* framework 0.27.0
* auth-mapper-* 0.2.0

# Index
1. [Requirements](#requirements)
2. [Installation](#installation)
3. [Before first run](#before-first-run)
    * [Protocol and app dir config](#protocol-and-app-dir-config)
    * [Router config](#router-config)
4. [Url rewriting](#url-rewriting)
    * [Apache Virtual Host config for mod_rewrite](#apache-virtual-host-config-for-mod_rewrite)
    * [Apache .htaccess config for mod_rewrite](#apache-htaccess-config-for-mod_rewrite)
    * [Nginx](#nginx)
5. [Dot env file](#dot-env-file)

# Requirements
* App need [linna/framework](https://github.com/linna/framework), please read the
[changelog](https://github.com/linna/app/blob/b0.16.0/CHANGELOG.md) for know with 
which version the app works.
* PHP 7.4 or higher.

> **Note:** App was only tested under Linux with Apache web server with default php.ini

# Installation
> **Note:** Consider use of sudo command if need administrator privileges and don't
forget to set proper folder permissions

With [composer](https://getcomposer.org/)
```Shell
cd /var/www/html
mkdir app
composer create-project --prefer-dist linna/app app
```
Where `app` is directory under webserver document root ex. `/var/www/html/app`

After, run composer [dump-autoload](https://getcomposer.org/doc/03-cli.md#dump-autoload) for optimize file autoloading
```Shell
composer dump-autoload --optimize
```

# Before first run
Change config in `config.php` file placed in `/var/www/html/app/config` directory.

## Protocol and app dir config
```php
$options = [

    'app' => [
        //protocol utilized [http://, https://]
        //default value set automatically
        'protocol'     => REQUEST_SCHEME.'://',
        //folder of the app, if app isn't in the web server root add a
        //directory (/app, /other/app) else insert a / (slash) as value
        //default value [/app]
        'subFolder'    => '/app',
        //public folder of the app, starting from web server root
        //default value [/app/public]
        'publicFolder' => '/app/public',
        //.env file position, remember to add ../ if don't use an absolute path
        'envFile'      => '../.env',
        //name of the fallback route, indicate the path when router return a NullRoute
        //default /error/404
        'onNullRoute'  => '/error/404'
    ],

    //other options
];
```

## Router config
```php
$options = [

    //other options

    'router' => [
        //must be equal to app.subFolder, it represents the part of the path
        //that the router ignore when check a route. Example '/app/user/delete/5'
        //become '/user/delete/5' where the router subtract the basePath
        //default [/app]
        'basePath'             => '/app',
        //url rewriting
        //default [true]
        'rewriteMode'          => true,
        //part of the url that the router ignore when url rewriting is off
        'rewriteModeOffRouter' => '/index.php?',
    ],

    //other options
];
```

# Url rewriting
If you enable the option of the router named `rewriteMode` in `config.php`,
need to configure your virtual host/server block.

## Apache Virtual Host config for mod_rewrite
For Apache VirtualHost config please see:  
[http://httpd.apache.org/docs/current/vhosts/](http://httpd.apache.org/docs/current/vhosts/)  
For Apache mod_rewrite config please see:  
[https://httpd.apache.org/docs/current/rewrite/](https://httpd.apache.org/docs/current/rewrite/)  
```ApacheConf
<VirtualHost *:80>

    # Other virtual host directives.

    <Directory /var/www/html/app>
        RewriteEngine on
        # Route to /app/public
        RewriteRule ^(.*)  public/$1 [L]
    </Directory>

    <Directory /var/www/html/app/public>
        # Necessary to prevent problems when using a controller named "index" and having a root index.php
        # more here: http://httpd.apache.org/docs/current/content-negotiation.html
        Options -MultiViews

        # Activates URL rewriting (like myproject.com/controller/action/1/2/3)
        RewriteEngine On

        # Prevent people from looking directly into folders
        Options -Indexes

        # If the following conditions are true, then rewrite the URL:
        # If the requested filename is not a directory,
        RewriteCond %{REQUEST_FILENAME} !-d
        # and if the requested filename is not a regular file that exists,
        RewriteCond %{REQUEST_FILENAME} !-f
        # and if the requested filename is not a symbolic link,
        RewriteCond %{REQUEST_FILENAME} !-l

        # then rewrite the URL in the following way:
        # Take the whole request filename and provide it as the value of a
        # "url" query parameter to index.php. Append any query string from
        # the original URL as further query parameters (QSA), and stop
        # processing (L).
        # https://httpd.apache.org/docs/current/rewrite/flags.html#flag_qsa
        # https://httpd.apache.org/docs/current/rewrite/flags.html#flag_l
        RewriteRule ^(.+)$ index.php [QSA,L]
    </Directory>

    # Other virtual host directives.

</VirtualHost>
```

## Apache .htaccess config for mod_rewrite
If you haven't access to your apache virtual host configuration,
you can add .htaccess files to the app for enable mod_rewrite.  

Create `.htaccess` file in `app/` directory with this content:
```ApacheConf
RewriteEngine on
# Route to /app/public
RewriteRule ^(.*)  public/$1 [L]
```

Create `.htaccess` file in `app/public/` directory with this content:
```ApacheConf
# Necessary to prevent problems when using a controller named "index" and having a root index.php
# more here: http://httpd.apache.org/docs/current/content-negotiation.html
Options -MultiViews

# Activates URL rewriting (like myproject.com/controller/action/1/2/3)
RewriteEngine On

# Prevent people from looking directly into folders
Options -Indexes

# If the following conditions are true, then rewrite the URL:
# If the requested filename is not a directory,
RewriteCond %{REQUEST_FILENAME} !-d
# and if the requested filename is not a regular file that exists,
RewriteCond %{REQUEST_FILENAME} !-f
# and if the requested filename is not a symbolic link,
RewriteCond %{REQUEST_FILENAME} !-l

# then rewrite the URL in the following way:
# Take the whole request filename and provide it as the value of a
# "url" query parameter to index.php. Append any query string from
# the original URL as further query parameters (QSA), and stop
# processing (L).
# https://httpd.apache.org/docs/current/rewrite/flags.html#flag_qsa
# https://httpd.apache.org/docs/current/rewrite/flags.html#flag_l
RewriteRule ^(.+)$ index.php [QSA,L]
```

## Nginx 
For Nginx Server Blocks config please see:  
[https://www.nginx.com/resources/wiki/start/topics/examples/server_blocks/](https://www.nginx.com/resources/wiki/start/topics/examples/server_blocks/)  

Setting url rewrite with Nginx is simpler than Apache counterpart, 
add `try_files $uri $uri/ /index.php?$args;` to `location` block:
```Nginx
server {

    # Other directives
    
    location / {
        # Url rewrite
        # Add line blow to location block for enable url rewriting
        try_files $uri $uri/ /index.php?$args;
    }

    # Other directives
}
```

# Dot env file
With `composer` installation, a `.env` file is created into `app` root directory 
and it could be used for declaring default environment variables.

`.env` file content look like this:
```
# Session
session.name   = linna_session
session.expire = 1800

## Pdo Mysql
pdo_mysql.user     = root
pdo_mysql.password =

## Mysqli
#mysqli.host     = 127.0.0.1
#mysqli.user     = root
#mysqli.password =
#mysqli.database = linna_db
#mysqli.port     = 3306

## MongoDB
#mongo_db.uri = mongodb://localhost:27017

## Memcached
#memcached.host = localhost
#memcached.port = 11211
```

`.env` file valid keys:
```
session.name
session.expire

pdo_mysql.dsn
pdo_mysql.user
pdo_mysql.password

pdo_pgsql.dsn
pdo_pgsql.user
pdo_pgsql.password

mysqli.host
mysqli.user
mysqli.password
mysqli.database
mysqli.port

mongo_db.uri

memcached.host
memcached.port
```

Values declared in the file will overwrite `config.php` values.

Position of `.env` file could be changed editing `envFile` value.
```php
$options = [

    'app' => [
        //other app options
        'envFile'           => '../.env'
    ],

    //other options
];
```

If you do not want use `.env` file can delete it.


CHANGELOG.md

# Linna App Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] [v0.17.0](https://github.com/linna/app/compare/v0.16.0...v0.17.0)


## [v0.16.1](https://github.com/linna/app/compare/v0.16.0...v0.16.1) - 2022-10-03

### Changed
* dependencies updated

## [v0.16.0](https://github.com/linna/app/compare/v0.15.0...v0.16.0) - 2022-09-17

### Added
* `Linna\App\Models\ErrorModel` class
* `Linna\App\Helper\RouteHelper` class, permit to declare routes directly in controller as annotation 
* option `onNullRoute` of `app` section with default value `/error/404` in `config/config.php`
* option `protocol` of `app` section now is set automatically
* `config.local.php` file for develop pourposes, it has priority on `config.php`

### Fixed
* namespace convention, all under `Linna\App` namespace
* `Linna\App\Helper\DotEnv` doesn't override configuration

### Changed
* require PHP 8.1
* require [linna/framework v0.27.0](https://github.com/linna/framework/releases/tag/v0.27.0)
* Error page system.
* `Linna\App\Controllers\E404Controller` renamed to `App\Controllers\ErrorController`
* `Linna\App\Views\E404View` renamed to `App\Views\ErrorView`

## [v0.15.0](https://github.com/linna/app/compare/v0.14.0...v0.15.0) - 2019-08-18

### Changed
* require PHP 7.2
* require [linna/framework v0.26.0](https://github.com/linna/framework/releases/tag/v0.26.0)
* `URL_STYLE` constant defined in `/public/index.php` renamed to `URL_PUBLIC`

### Added
* `App\Helper\AppDotEnv->registerKey()` method, add a new valid key for .env file
* `App\Helper\AppDotEnv->registerKeys()` method, add a new set of valid keys for .env file
* `App\Templates\RawTemplate` class
* `App\Templates\HtmlTemplate->__construct()` 3rd argument as `string $jsDir`
* `Linna\Tests\HtmlTemplateTest` class
* `Linna\Tests\JsonTemplateTest` class
* `Linna\Tests\NullTemplateTest` class
* `Linna\Tests\RawTemplateTest` class

### Fixed
* `App\Templates\HtmlTemplate` load js files from css folder

## [v0.14.0](https://github.com/linna/app/compare/v0.13.0...v0.14.0) - 2019-01-13

### Added
* `App\Controllers\NullController` class
* `App\Helper\AppDotEnv;` class
* `App\Templates\NullTemplate` class
* `App\Views\NullView` class
* Dotenv section in `public/index.php`
* option `envFile` to `app` section in `config/config.php`

### Changed
* require [linna/framework v0.25.0](https://github.com/linna/framework/releases/tag/v0.25.0)
* `composer.json` autoload section
* all App classes are loaded with composer autoload

### Removed
* Autoloader section from `public/index.php`
* `bin/compile-routes` script
* `config/routes.compiled.php` file
* option `compiledRoutes` to `app` section in `config/config.php`

## [v0.13.0](https://github.com/linna/app/compare/v0.12.0...v0.13.0) - 2018-09-01

### Changed
* require [linna/framework v0.24.0](https://github.com/linna/framework/releases/tag/v0.24.0)
* minimun PHP version: 7.1
* E404 route now responds with 404 status code

### Fixed
* compiled routes bug in `index.php`

## [v0.12.0](https://github.com/linna/app/compare/v0.11.0...v0.12.0) - 2017-10-24

### Changed
* require [linna/framework v0.22.0](https://github.com/linna/framework/releases/tag/v0.22.0)

## [v0.11.0](https://github.com/linna/app/compare/v0.10.1...v0.11.0) - 2017-10-04

### Changed
* `App\Models\E404Model` renamed to `App\Models\NullModel`, use it when page do not require elaborated data
* `App\Templates\HtmlTemplate->__construct()` now require as parameters template and css directory
* require [linna/framework v0.21.0](https://github.com/linna/framework/releases/tag/v0.21.0)
* option `urlProtocol` renamed to `protocol`
* option `urlSubFolder` renamed to `subFolder`
* option `urlPublicFolder` renamed to `publicFolder`
* option `useCompiledRoutes` renamed to `compiledRoutes`

### Added
* `App\Templates\HtmlTemplate` section to `config/injections.php`

## [v0.10.1](https://github.com/linna/app/compare/v0.10.0...v0.10.1) - 2017-07-25

### Changed
* `config/config.php` option `useCompiledRoutes` default changed to `false`

### Added
* `config/config.php` option `rewriteModeOffRouter` in `router` section

### Fixed
* app working in rewrite mode off

## [v0.10.0](https://github.com/linna/app/compare/v0.9.0...v0.10.0) - 2017-07-17

### Added
* `bin/compile-routes` script for export routes as php array
* possibility to declare routes as `Linna\Http\RouteCollection` and export it as php array
* `favicon.png` in `/public` directory
* `robots.txt` in `/public` directory

### Changed
* `declare(strict_types=1);` added where missing
* `.htaccess` configurations moved to virtual host config file
* `README.md` virtual host rewrite mod config added
* require [linna/framework v0.20.0](https://github.com/linna/framework/releases/tag/v0.20.0)

### Fixed
* `CHANGELOG.md` links url
* file permissions

### Removed
* `.htaccess` in `/` directory
* `.htaccess` in `/public` directory

## [v0.9.0](https://github.com/linna/app/v0.8.0...v0.9.0) - 2017-06-24

### Changed
* require [linna/framework v0.19.0](https://github.com/linna/framework/releases/tag/v0.19.0)

## [v0.8.0](https://github.com/linna/app/compare/v0.7.0...v0.8.0) - 2017-06-01

### Removed
* all files for login and user
* EnhancedUserMapper.php from `src/Mappers` directory
* PermissionMapper.php from `src/Mappers` directory
* RoleMapper.php from `src/Mappers` directory

### Changed
* app theme and templates
* templates that implements `Linna\Mvc\TemplateInterface` now contain `public function getOutput() : string` instead of `public function output()`
* option default for `urlSubFolder` changed from `/app/` to `/app`
* option default for `urlPublicFolder` changed from `public` to `/app/public`
* `Linna\Mvc\FrontController` usage changed on `public/index.php`

config/config.php
<?php

/**
 * Linna App.
 *
 * @author Sebastian Rapetti <sebastian.rapetti@tim.it>
 * @copyright (c) 2020, Sebastian Rapetti
 * @license http://opensource.org/licenses/MIT MIT License
 */
declare(strict_types=1);

return [

    'app' => [
        //protocol utilized [http://, https://]
        //default value set automatically
        'protocol'     => REQUEST_SCHEME.'://',
        //folder of the app, if app isn't in the web server root add a
        //directory (/app, /other/app) else insert a / (slash) as value
        //default value [/app]
        'subFolder'    => '/',
        //public folder of the app, starting from web server root
        //default value [/app/public]
        'publicFolder' => '/public',
        //.env file position, remember to add ../ if don't use an absolute path
        'envFile'      => '../.env',
        //name of the fallback route, indicate the path when router return a NullRoute
        //default /error/404
        'onNullRoute'  => '/error/404'
    ],

    'modules' => [
        //'10-storage-mysql',
        //'10-storage-pgsql',
        //'20-session-mysql',
        //'20-session-pgsql',
        '20-session',
        '30-router'
    ],

    'session' => [
        //session name
        //default [linna_session]
        'name'           => 'linna_session',
        //session expire time in seconds
        //default [1800]
        'expire'         => 1800,
        //cookie scope, domain and path where the cookie will be sent.
        //for cookieDomain the value of constant URL_DOMAIN is obtenined
        //from $_SERVER['HTTP_HOST'], for cookie path the value is equal to
        //app.subFolder value
        'cookieDomain'   => URL_DOMAIN, //do not change here
        'cookiePath'     => '/',
        //cookie will sent only with https requests
        //default [false]
        'cookieSecure'   => false,
        //when set to true, cookie is inaccessible from Javascript
        //default [true]
        'cookieHttpOnly' => true,
    ],

    'router' => [
        //must be equal to app.subFolder, it represents the part of the path
        //that the router ignore when check a route. Example '/app/user/delete/5'
        //become '/user/delete/5' where the router subtract the basePath
        //default [/app]
        'basePath'             => '/',
        //url rewriting
        //default [true]
        'rewriteMode'          => true,
        //part of the url that the router ignore when url rewriting is off
        'rewriteModeFalseEntryPoint' => '/index.php?',
    ],

    'pdo_pgsql' => [
        'dsn'      => 'pgsql:dbname=linna_db;host=localhost',
        'user'     => 'postgres',
        'password' => '',
        'options'  => [
            \PDO::ATTR_DEFAULT_FETCH_MODE => \PDO::FETCH_OBJ,
            \PDO::ATTR_ERRMODE            => \PDO::ERRMODE_EXCEPTION,
            \PDO::ATTR_PERSISTENT         => false,
        ],
    ],

    'pdo_mysql' => [
        'dsn'      => 'mysql:host=localhost;dbname=linna_db;charset=utf8mb4',
        'user'     => 'root',
        'password' => '',
        'options'  => [
            \PDO::ATTR_DEFAULT_FETCH_MODE => \PDO::FETCH_OBJ,
            \PDO::ATTR_ERRMODE            => \PDO::ERRMODE_EXCEPTION,
            \PDO::ATTR_PERSISTENT         => false,
            \PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci',
        ],
    ],

    'mysqli' => [
        'host'     => '127.0.0.1',
        'user'     => '',
        'password' => '',
        'database' => 'linna_db',
        'port'     => 3306,
    ],

    'mongo_db' => [
        'uri'           => 'mongodb://localhost:27017',
        'uriOptions'    => [],
        'driverOptions' => [],
    ],

    'memcached' => [
        'host' => 'localhost',
        'port' => 11211,
    ],

    'password_algo' => PASSWORD_DEFAULT,

    'password_options' => [
        'cost' => 11
    ],
];
